### DISCUSSION TITLE
Customised elements

### DISCUSSION THREAD

**Discussion by zia-javanbakht** [idaholab/moose#26618] [Q&A Modules: Solid mechanics]
I am trying to derive the week form of three coupled equilibrium equations for a 3-layer plate. I can do the FE formulation using additional DoFs, which requires a new type of element. How can I do this? Thanks.

**Comment by GiudGiud** [comment]
Hello

Making custom elements is more of a libmesh change than something we can do in MOOSE.
What type of new element do you need?

Guillaume

**Reply by zia-javanbakht** [reply]
Hi,

Thank you for your reply. I am modelling a three-layer sandwich structure. The elements can be Lagrange ones, but I the DoFs are a combination of displacements and rotations of the three layers. For instance, the first DoF is the average of deflections of the top and bottom layers, the second DoF is the average of their rotations. I am not sure how to address this.

As a side question, I have used the principle of virtual work and obtained the virtual internal and external works. I am also struggling to take this form to the standard kernel form that MOOSE requires. 

Thank you very much for your help.

kind regards,
Zia

**Reply by GiudGiud** [reply]
Could you use multiple variables to host each dof?
We for example can have X and Y components of the displacement in two variables, and then there are two DOFs at each node (one for X, one for Y)

**Comment by zia-javanbakht** [comment]
first DoF = average of the first components of deformation for layers 1 and 3
second DoF = average of the second components of deformation for layers 1 and 3
third Dof = difference of the first components of deformation for layers 1 and 3
...

![Screenshot 2024-01-25 084203](https://github.com/idaholab/moose/assets/67005474/04f29a0f-3f97-4804-8095-79ad642954aa)

**Reply by GiudGiud** [reply]
Should you make these averages variables that are only defined in layer 2 then?
These dofs are supposed to live on layer 2 right?

**Reply by zia-javanbakht** [reply]
Basically, the three layers are modelled as one layer with all the DoFs on it. I can send you my paper if it helps.

**Reply by GiudGiud** [reply]
I think you could a one layer mesh with the 9 variables defined in a as in the figure above.

We can solve problems with multiple variables. 
If some of the variables are simple averages , you can use auxiliary variables to take them out of the nonlinear solver

**Reply by zia-javanbakht** [reply]
The virtual work formulation is based on these averages so the DoFs must be these variables. Is there any examples available for something like this? Also do you know how to reformulate the virtual work expression into weak form suitable for MOOSE? Thanks.

**Reply by GiudGiud** [reply]
I dont know about the virtual work. @gambka @cbolisetti  ?
`the DoFs must be these variables` seems like you should have a variable field for each of them.

**Reply by zia-javanbakht** [reply]
Thank you very much for answering my questions. Do I need to do anything specific for a "variable field"?

**Reply by GiudGiud** [reply]
A variable in moose is a field that is defined over part or all of the mesh. 

There are variables that have DoFS on nodes (for example Lagrange) and others that have dofs at various locations on elements (for example monomial or Bernstein)

Guillaume
